warehouse/admin/views/prohibited_email_domains.py

```diff
@@ -0,0 +1,131 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from paginate_sqlalchemy import SqlalchemyOrmPage as SQLAlchemyORMPage
+from pyramid.httpexceptions import HTTPBadRequest, HTTPSeeOther
+from pyramid.view import view_config
+from sqlalchemy import func
+from tldextract import extract
+
+from warehouse.accounts.models import ProhibitedEmailDomain
+from warehouse.authnz import Permissions
+from warehouse.utils.paginate import paginate_url_factory
+
+
+@view_config(
+    route_name="admin.prohibited_email_domains.list",
+    renderer="admin/prohibited_email_domains/list.html",
+    permission=Permissions.AdminProhibitedEmailDomainsRead,
+    request_method="GET",
+    uses_session=True,
+)
+def prohibited_email_domains(request):
+    q = request.params.get("q")
+
+    try:
+        page_num = int(request.params.get("page", 1))
+    except ValueError:
+        raise HTTPBadRequest("'page' must be an integer.") from None
+
+    prohibited_email_domains_query = request.db.query(ProhibitedEmailDomain).order_by(
+        ProhibitedEmailDomain.created.desc()
+    )
+
+    if q:
+        prohibited_email_domains_query = prohibited_email_domains_query.filter(
+            ProhibitedEmailDomain.domain.ilike(q)
+        )
+
+    prohibited_email_domains = SQLAlchemyORMPage(
+        prohibited_email_domains_query,
+        page=page_num,
+        items_per_page=25,
+        url_maker=paginate_url_factory(request),
+    )
+
+    return {"prohibited_email_domains": prohibited_email_domains, "query": q}
+
+
+@view_config(
+    route_name="admin.prohibited_email_domains.add",
+    permission=Permissions.AdminProhibitedEmailDomainsWrite,
+    request_method="POST",
+    uses_session=True,
+    require_methods=False,
+)
+def add_prohibited_email_domain(request):
+    email_domain = request.POST.get("email_domain")
+    if not email_domain:
+        request.session.flash("Email domain is required.", queue="error")
+        raise HTTPSeeOther(request.route_path("admin.prohibited_email_domains.list"))
+    # validate that the domain is valid
+    if not extract(email_domain).registered_domain:
+        request.session.flash(f"Invalid domain name '{email_domain}'", queue="error")
+        raise HTTPSeeOther(request.route_path("admin.prohibited_email_domains.list"))
+    # make sure we don't have a duplicate entry
+    if (
+        request.db.query(func.count(ProhibitedEmailDomain.id))
+        .filter(ProhibitedEmailDomain.domain == email_domain)
+        .scalar()
+        > 0
```

Comment:
My understanding is something like this is generally preferred for testing presence of a row:

```suggestion
        request.db.session.query(
            exists().where(ProhibitedEmailDomain.domain == email_domain)
        ).scalar()
```
Is there any reason to prefer one over the other?

Reply:
Changed in 5e519f66b9802f0ef9e1f9183dcd44a9e890d2a5 - syntax is a little more explicit.

The real answer is that exists subquery is often faster than a count (aggregate) query - since exists stops at the first result found, and count wants to count all.

For small tables like this one, it's unlikely to ever grow to the scale of being a material issue, but thanks for calling it out!